onnxruntime/core/providers/cpu/nn/string_split.cc

```diff
@@ -0,0 +1,112 @@
+#include "string_split.h"
+#include "core/common/common.h"
+#include <algorithm>
+namespace onnxruntime {
+
+ONNX_CPU_OPERATOR_KERNEL(
+    StringSplit,
+    20,
+    KernelDefBuilder()
+        .TypeConstraint("T1", DataTypeImpl::GetTensorType<std::string>())
+        .TypeConstraint("T2", DataTypeImpl::GetTensorType<std::string>())
+        .TypeConstraint("T3", DataTypeImpl::GetTensorType<int64_t>()),
+    StringSplit);
+
+int64_t countSubstrings(std::string_view str, std::string_view substr) {
+  if (str.empty()) {
+    return 0;
+  }
+  if (substr.empty()) {
+    // Count consecutive whitespace as one delimiter
+    int64_t count = 1;
+    size_t pos = str.find_first_not_of(" ");
+    while (pos != std::string::npos) {
+      ++count;
+      pos = str.find_first_not_of(" ", str.find_first_of(" ", pos));
+    }
+    return count;
+  } else {
+    int64_t count = 1;
+    size_t pos = str.find(substr);
+    while (pos != std::string::npos) {
+      ++count;
+      pos = str.find(substr, pos + substr.length());
+    }
+    return count;
+  }
+}
+
+int64_t fillSubstrings(std::string_view str, std::string_view delimiter, gsl::span<std::string> output, int64_t output_index, size_t max_tokens) {
+  if (str.empty()) {
+    return 0;
+  }
+  if (delimiter.empty()) {
+    // Count consecutive whitespace as one delimiter. Preceding and trailing whitespace is meant to be ignored.
+    size_t pos = str.find_first_not_of(" ");
+    size_t token_index = 0;
+    while (token_index < max_tokens && pos != std::string::npos) {
+      auto next_pos = token_index == max_tokens - 1 ? std::string::npos : str.find_first_of(" ", pos);
+      output[output_index + token_index] = str.substr(pos, next_pos - pos);
+      ++token_index;
```

Comment:
Would it better to supply output span specifically for this string instead of a global one when the index has to re-calculate on every iteration?

Reply:
It is also worth measuring for large tensors (and they will be large) if scanning the input 3 times (count + search again + substr) is slow. One could memorize the substring positions on the first scan, and then re-use it simply to substr() on the second pass.

Reply:
I don't quite see how there are 3 scans of the input? 
The first one is a count over the number of tokens (https://github.com/microsoft/onnxruntime/pull/18016/files#diff-e2b1d01bf1079e9b27f3288762e6896c104fb72c29225f6abfad52414de42f69R97-R99) (used to calculate the output shape).

The second is to produce the substrings + split count (https://github.com/microsoft/onnxruntime/pull/18016/files#diff-e2b1d01bf1079e9b27f3288762e6896c104fb72c29225f6abfad52414de42f69R115-R117).

Am I missing something?

Reply:
Basically, we could reduce data reading times by one.
If `CountSubstrings` did not only count, but also record std::string_views of the substrings, then we either simplify the second function or get rid of it.
The only thing we would have to do is to copy data from recorded string views to the output and pad it.

Reply:
I see, nice idea. I've implemented something to this effect.